pkg/deploy/controller/deploymentconfig/deploymentconfig_controller.go

```diff
@@ -251,6 +217,147 @@ func (c *DeploymentConfigController) Handle(config *deployapi.DeploymentConfig)
+			updatedDeployment, err = c.rn.ReplicationControllers(copied.Namespace).Update(copied)
+			return err
+		})
+		if err != nil {
+			c.recorder.Eventf(config, v1.EventTypeWarning, "DeploymentCancellationFailed", "Failed to cancel deployment %q superceded by version %d: %s", deployment.Name, config.Status.LatestVersion, err)
+			return nil
+		}
+		if updatedDeployment != nil {
+			// replace the current deployment with the updated copy so that a future update has a chance at working
+			existingDeployments[i] = updatedDeployment
+			c.recorder.Eventf(config, v1.EventTypeNormal, "DeploymentCancelled", "Cancelled deployment %q superceded by version %d", deployment.Name, config.Status.LatestVersion)
+		}
+	}
+
+	// Wait for deployment cancellations before reconciling or creating a new
+	// deployment to avoid competing with existing deployment processes.
+	if awaitingCancellations {
+		c.recorder.Eventf(config, v1.EventTypeNormal, "DeploymentAwaitingCancellation", "Deployment of version %d awaiting cancellation of older running deployments", config.Status.LatestVersion)
+		return fmt.Errorf("found previous inflight deployment for %s - requeuing", deployutil.LabelForDeploymentConfig(config))
+	}
+
+	return nil
+}
+
+// triggerActivated indicates whether we should proceed with new rollout as one of the
+// triggers were activated (config change or image change). The first bool indicates that
+// the triggers are active and second indicates if we should skip the rollout because we
+// are waiting for the trigger to complete update (waiting for image for example).
+func triggerActivated(config *deployapi.DeploymentConfig, latestIsDeployed bool, latestDeployment *v1.ReplicationController) (bool, bool) {
+	if config.Spec.Paused {
+		return false, false
+	}
+	imageTrigger := deployutil.HasImageChangeTrigger(config)
+	configTrigger := deployutil.HasChangeTrigger(config)
+	hasTrigger := imageTrigger || configTrigger
+
+	// no-op when no triggers are defined.
+	if !hasTrigger {
+		return false, false
+	}
+
+	// Handle initial rollouts
+	if deployutil.IsInitialDeployment(config) {
+		hasAvailableImages := deployutil.HasLastTriggeredImage(config)
+		// Rollout if image change triggers are present and all triggers have their images
+		// available.
+		if imageTrigger && hasAvailableImages {
+			glog.Infof("Rolling out initial deployment for %s/%s as it have its images available", config.Namespace, config.Name)
+			// TODO: This is not really config change but to record image change we will need to
+			// know the image that triggered, but in some cases this might be multiple images.
+			deployutil.RecordConfigChangeCause(config)
+			return true, false
+		}
+		// Skip rollout if image change triggers are present but the images are not ready yet.
+		if imageTrigger && !hasAvailableImages {
+			return false, true
+		}
+		// Rollout if we only have config change trigger.
+		if configTrigger && !imageTrigger {
```

Comment:
Testing for `!imageTrigger` seems redundant since it is already ruled out by the code above. But generally this logic seems slightly confusing since it tests conditions twice. Could we do:
```go
if imageTrigger {
    if hasAvailableImages {
        return
    }
   return
}

if configTrigger {
    return
}